contracts-periphery/src/UmbraBatchSend.sol

```diff
@@ -24,69 +27,133 @@ contract UmbraBatchSend {
+    for (uint256 j = 0; j < summaryLength; ) {
+      IERC20 token = IERC20(address(_transferSummary[j].tokenAddr));
+
+      if (TotalTransferAmountPerToken[_transferSummary[j].tokenAddr] != _transferSummary[j].amount)
+        revert TransferAmmountMismatch();
+      SafeERC20.safeTransferFrom(token, msg.sender, address(this), _transferSummary[j].amount);
+
+      unchecked {
+        j++;
+      }
+    }
+    _batchSendTokens(_tollCommitment, _params, _transferSummary);
+
     emit BatchSendExecuted(msg.sender);
   }
 
   function batchSend(
     uint256 _tollCommitment,
     SendEth[] calldata _ethParams,
-    SendToken[] calldata _tokenParams
+    SendToken[] calldata _tokenParams,
+    TransferSummary[] calldata _transferSummary
   ) external payable {
     uint256 _totalAmount = _sumValueFromEthSends(_tollCommitment, _ethParams) + (_tollCommitment * _tokenParams.length);
-    if(msg.value != _totalAmount) revert ValueMismatch();
+    if (msg.value != _totalAmount) revert ValueMismatch();
 
     _batchSendEth(_tollCommitment, _ethParams);
-    _batchSendTokens(_tollCommitment, _tokenParams);
+    batchSendTokens(_tollCommitment, _tokenParams, _transferSummary);
     emit BatchSendExecuted(msg.sender);
   }
 
   function _batchSendEth(uint256 _tollCommitment, SendEth[] calldata _params) internal {
-    for (uint256 i = 0; i < _params.length; i++) {
-      umbra.sendEth{value: _params[i].amount + _tollCommitment}(_params[i].receiver, _tollCommitment, _params[i].pkx, _params[i].ciphertext);
+    uint256 length = _params.length;
+    for (uint256 i = 0; i < length; ) {
+      umbra.sendEth{value: _params[i].amount + _tollCommitment}(
+        _params[i].receiver,
+        _tollCommitment,
+        _params[i].pkx,
+        _params[i].ciphertext
+      );
+      unchecked {
+        i++;
+      }
     }
   }
 
-  function _batchSendTokens(uint256 _tollCommitment, SendToken[] calldata _params) internal {
-    for (uint256 i = 0; i < _params.length; i++) {
-      IERC20 token = IERC20(address(_params[i].tokenAddr));
-
-      SafeERC20.safeTransferFrom(token, msg.sender, address(this), _params[i].amount);
-
-      if (token.allowance(address(this), address(umbra)) == 0) {
-        SafeERC20.safeApprove(token, address(umbra), type(uint256).max);
-      }
-
+  function _batchSendTokens(
+    uint256 _tollCommitment,
+    SendToken[] calldata _params,
+    TransferSummary[] calldata _transferSummary
+  ) internal {
+    uint256 length = _params.length;
+    for (uint256 i = 0; i < length; ) {
       umbra.sendToken{value: _tollCommitment}(
         _params[i].receiver,
         _params[i].tokenAddr,
         _params[i].amount,
         _params[i].pkx,
         _params[i].ciphertext
       );
+      unchecked {
+        i++;
+      }
+    }
+
+    uint256 summaryLength = _transferSummary.length;
+    for (uint256 j = 0; j < summaryLength; ) {
+      IERC20 token = IERC20(address(_transferSummary[j].tokenAddr));
```

Comment:
This local variable seems to be declared but not used

Reply:
Removed it! https://github.com/ScopeLift/umbra-protocol/blob/b156045ea6b0d75c49dbb6500e5e9e6835131723/contracts-periphery/src/UmbraBatchSend.sol#L128-L132